neoemf-core/src/main/java/fr/inria/atlanmod/neoemf/core/internal/ContentsListIterator.java

```diff
@@ -0,0 +1,484 @@
+/*
+ * Copyright (c) 2013-2018 Atlanmod, Inria, LS2N, and IMT Nantes.
+ *
+ * All rights reserved. This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License v2.0 which accompanies
+ * this distribution, and is available at https://www.eclipse.org/legal/epl-2.0/
+ */
+
+package fr.inria.atlanmod.neoemf.core.internal;
+
+import fr.inria.atlanmod.neoemf.core.PersistentEObject;
+import fr.inria.atlanmod.neoemf.util.EObjects;
+
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.EStructuralFeature;
+import org.eclipse.emf.ecore.util.EContentsEList;
+import org.eclipse.emf.ecore.util.FeatureMap;
+import org.eclipse.emf.ecore.util.FeatureMapUtil;
+import org.eclipse.emf.ecore.util.InternalEList;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.NoSuchElementException;
+
+import javax.annotation.Nonnegative;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.annotation.ParametersAreNonnullByDefault;
+
+import static fr.inria.atlanmod.commons.Preconditions.checkState;
+import static java.util.Objects.isNull;
+import static java.util.Objects.nonNull;
+
+/**
+ * A read-only {@link ListIterator} that iterates over all the content of a {@link PersistentEObject}.
+ *
+ * @param <E> the type of elements in this iterator
+ *
+ * @see ContentsList
+ * @see PersistentEObject#eContents()
+ * @see PersistentEObject#eAllContents()
+ */
+@ParametersAreNonnullByDefault
+class ContentsListIterator<E extends EObject> implements EContentsEList.FeatureListIterator<E>, EContentsEList.Filterable {
+
+    /**
+     * The instance of an empty {@code ContentsListIterator}.
+     */
+    @Nonnull
+    private static final ContentsListIterator<?> EMPTY = new EmptyContentsListIterator<>();
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Remove this reference to "EmptyContentsListIterator". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2390)